### PR TITLE
Add `energy_ensemble` as a new standard output for atomistic models

### PR DESCRIPTION
**Pull request by frostedoyster**
This PR implements ensembles of energies as a new standard output for atomistic models (#650). It also makes the necessary changes to the ASE calculator to make this property (which is not standard in ASE) available to the user. In addition, the dipole is also registered as a new property and added to the ASE calculator

TODO: actual conversion factors for dipoles

# Contributor (creator of pull-request) checklist

 - [ ] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [x] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
 📚 [Download documentation preview for this pull-request](https://nightly.link/lab-cosmo/metatensor/actions/artifacts/1719201793.zip)

### CODE REVIEW

**Comment by frostedoyster** [comment]
Regarding the tests, it seems that `energy` and `energy_ensemble` are being ordered differently on Windows x)

File: metatensor-torch/src/atomistic/model.cpp [hunk 1]
```diff
@@ -1032,6 +1034,29 @@ static std::unordered_map<std::string, Quantity> KNOWN_QUANTITIES = {
         {"J", "Joule"},
         {"Ry", "Rydberg"},
     }}},
+    {"energy_ensemble", Quantity{/* name */ "energy", /* baseline */ "eV", {
```
Comment (Luthaf) [inline metatensor-torch/src/atomistic/model.cpp:1037]: This should not be needed, `energy_ensemble` should be able to re-use the `energy` quantity and associated units

File: metatensor-torch/src/atomistic/model.cpp [hunk 2]
```diff
@@ -1032,6 +1034,29 @@ static std::unordered_map<std::string, Quantity> KNOWN_QUANTITIES = {
...
+        {"Joule", 1.60218e-19},
+        {"Rydberg", 0.07349864435130857},
+    }, {
+        // alternative names
+        {"J", "Joule"},
+        {"Ry", "Rydberg"},
+    }}},
+    {"dipole", Quantity{/* name */ "dipole", /* baseline */ "D", {
+        {"Debye", 1.0},
+        {"Coulomb-meter", 1000.0},
```
Comment (Luthaf) [inline metatensor-torch/src/atomistic/model.cpp:1052]:
```suggestion
        {"Coulombmeter", 1000.0},
```

File: metatensor-torch/src/atomistic/model.cpp [hunk 3]
```diff
@@ -1032,6 +1034,29 @@ static std::unordered_map<std::string, Quantity> KNOWN_QUANTITIES = {
...
+        {"atomic units", 0.03674932247495664},
+    }, {
+        // alternative names
+        {"D", "Debye"},
+        {"C-m", "Coulomb-meter"},
```
Comment (Luthaf) [inline metatensor-torch/src/atomistic/model.cpp:1057]:
```suggestion
        {"Cm", "Coulombmeter"},
```

This seems to be the SI unit here

File: metatensor-torch/src/atomistic/model.cpp [hunk 4]
```diff
@@ -1032,6 +1034,29 @@ static std::unordered_map<std::string, Quantity> KNOWN_QUANTITIES = {
...
+        {"Rydberg", 0.07349864435130857},
+    }, {
+        // alternative names
+        {"J", "Joule"},
+        {"Ry", "Rydberg"},
+    }}},
+    {"dipole", Quantity{/* name */ "dipole", /* baseline */ "D", {
+        {"Debye", 1.0},
+        {"Coulomb-meter", 1000.0},
+        {"atomic units", 0.03674932247495664},
```
Comment (Luthaf) [inline metatensor-torch/src/atomistic/model.cpp:1053]: I'm not sure I love having a space inside the unit =/

**Review by Luthaf — COMMENTED** [review]
This is changing way too many things at the same time. Could you have one PR for energy-ensemble, one PR for dipole and one PR changing the ASE calculator? 

The new outputs need code to check that the data matches the specification in https://github.com/lab-cosmo/metatensor/blob/master/python/metatensor-torch/metatensor/torch/atomistic/outputs.py

File: python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py [hunk 5]
```diff
@@ -73,6 +74,12 @@ def __init__(
             running, defaults to False.
         :param device: torch device to use for the calculation. If ``None``, we will try
             the options in the model's ``supported_device`` in order.
+        :param properties_to_store: list of model outputs to store as results of the ASE
+            calculator at every step. This is useful when you want to store properties
+            that are not used in the propagation of the dynamics and/or are not standard
+            ASE properties ('energy', 'forces', 'stress', 'stresses', 'dipole',
+            'charges', 'magmom', 'magmoms', 'free_energy', 'energies'). These properties
+            will be available as ``atoms.calc.results['<stored_property>']``.
```
Comment (Luthaf) [inline python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py:82]: Is this typical in ASE? The idea for properties not part of the standard was to use the `run_model` function.

In general, this does not seems to be required for dipole, since ASE already supports them as property.

File: metatensor-torch/src/atomistic/model.cpp (hunk 3 above)
Comment (Luthaf) [inline metatensor-torch/src/atomistic/model.cpp:1057]: Should this be `C.m`? That's what https://en.wikipedia.org/wiki/Electric_dipole_moment uses

File: metatensor-torch/src/atomistic/model.cpp [hunk 6]
```diff
@@ -1032,6 +1034,29 @@ static std::unordered_map<std::string, Quantity> KNOWN_QUANTITIES = {
...
+        {"kcal/mol", 23.060548012069496},
+        {"kJ/mol", 96.48533288249877},
+        {"Joule", 1.60218e-19},
+        {"Rydberg", 0.07349864435130857},
+    }, {
+        // alternative names
+        {"J", "Joule"},
+        {"Ry", "Rydberg"},
+    }}},
+    {"dipole", Quantity{/* name */ "dipole", /* baseline */ "D", {
```
Comment (Luthaf) [inline metatensor-torch/src/atomistic/model.cpp:1050]: do we want to call this `electric_dipole`?

**Comment by frostedoyster** [comment]
Ok, I will open different PRs

**Comment by sirmarcel** [comment]
Thx Guillaume & Filippo. Feel free to re-assign me for ASE changes or ensemble stuff. 👍

File: docs/src/atomistic/outputs.rst [hunk 7]
```diff
@@ -119,3 +119,45 @@ The following gradients can be defined and requested with
     - ``["xyz_1", "xyz_2"]``
     - Both ``"xyz_1"`` and ``"xyz_2"`` have values ``[0, 1, 2]``, and correspond
       to the two axes of the 3x3 strain matrix :math:`\epsilon`.
+
+
+Energy ensemble
+^^^^^^^^^^^^^^^
+
+An ensemble of energies is associated with the ``"energy_ensemble"`` key in the
+model outputs, and must have the following metadata:
```
Comment (Luthaf) [inline docs/src/atomistic/outputs.rst:128]: I'm split whether we should duplicate the information that is identical to `energy` here, or say "keys: same as `energy`", "samples: same as `energy`", ... and only specify the bit that changes.

Do you have a preference?

File: python/metatensor-torch/metatensor/torch/atomistic/outputs.py [hunk 8]
```diff
@@ -172,3 +176,96 @@ def _check_energy(
...
+def _check_energy_ensemble(
+    systems: List[System],
+    request: ModelOutput,
+    selected_atoms: Optional[Labels],
+    energy_ensemble: TensorMap,
+):
+    """Check the "energy_ensemble" output metadata"""
+    if energy_ensemble.keys != Labels("_", torch.tensor([[0]])):
+        raise ValueError(
+            "invalid keys for 'energy' output: expected `Labels('_', [[0]])`"
```
Comment (Luthaf) [inline python/metatensor-torch/metatensor/torch/atomistic/outputs.py:190]:
```suggestion
            "invalid keys for 'energy_ensemble' output: expected `Labels('_', [[0]])`"
```

File: python/metatensor-torch/metatensor/torch/atomistic/outputs.py [hunk 9]
```diff
@@ -172,3 +176,96 @@ def _check_energy(
...
+        raise ValueError(
+            "invalid properties for 'energy_ensemble' output: expected a Labels "
+            "object with the name 'ensemble_member' and the values being the indices "
+            "of the ensemble members from 0 to the number of ensemble members minus "
+            "one."
+        )
+
+    if len(energy_ensemble_block.gradients()) != 0:
+        raise ValueError(
+            "invalid gradients for 'energy' output: no gradients should be present."
```
Comment (Luthaf) [inline python/metatensor-torch/metatensor/torch/atomistic/outputs.py:270]:
```suggestion
            "invalid gradients for 'energy_ensemble' output: no gradients should be present."
```

Comment (Luthaf) [reply python/metatensor-torch/metatensor/torch/atomistic/outputs.py:270]: Do we actually want to disallow the possibility to have forward gradients for `energy_ensemble`?

File: docs/src/atomistic/outputs.rst (hunk 7 above)
Comment (frostedoyster) [reply docs/src/atomistic/outputs.rst:128]: I would definitely prefer "same as energies"

File: python/metatensor-torch/metatensor/torch/atomistic/outputs.py (hunk 9 above)
Comment (frostedoyster) [reply python/metatensor-torch/metatensor/torch/atomistic/outputs.py:270]: Yeah... they make no sense... unless we give them a different meaning.
For example, the uncertainties on the forces are not the derivatives of the uncertainties on the energies

Comment (Luthaf) [reply python/metatensor-torch/metatensor/torch/atomistic/outputs.py:270]: I mean, the output is not called "energy_uncertainties" but "energy_ensemble", so the positions gradients would be an ensemble of forces. It would then be up to the engine to do what they want with it.

I'm also fine disallowing it for now, but then we should mention it in the standard.

File: docs/src/atomistic/outputs.rst (hunk 7 above)
Comment (Luthaf) [reply docs/src/atomistic/outputs.rst:128]: let's do this then

File: python/metatensor-torch/metatensor/torch/atomistic/outputs.py (hunk 9 above)
Comment (frostedoyster) [reply python/metatensor-torch/metatensor/torch/atomistic/outputs.py]: You're right, I got confused. The gradients for `energy_ensemble` are well-defined and it's a good idea to allow them as explicit gradients

File: docs/src/atomistic/outputs.rst (hunk 7 above)
Comment (frostedoyster) [reply docs/src/atomistic/outputs.rst:132]: Should be done

**Comment by Luthaf** [comment]
This looks good. Should we also have a test checking some example model that returns an ensemble of energies?
